### PR TITLE
Login UI

### LINKED ISSUE #108: Implement the UI for logging in/out
Requires an artifact from #99, but most of the frontend code and the tests can be implemented independently.

**Acceptance criteria**:
- the _Your Library_ page should contain the button to log in or log out
- the design of the authentication element should look identical to the proposed design:
  image
- unit tests should be added for the code

Note: you don't have to implement the authentication flow itself, that is, no need to handle responses from Supabase, store the user data or anything. Create a widget that will display either of the states based on the value of a prop like `currentUser` or something.

### PR DESCRIPTION
**Pull request by VanishMax**
When implemented, closes #108 

Tried to follow the FSD as much as possible :)

### CODE REVIEW

File: src/lib/entities/user/index.ts [hunk 1]
```diff
@@ -0,0 +1,2 @@
+export type { User } from './model/types';
+export { user } from './model/store';
```
Comment (illright) [inline src/lib/entities/user/index.ts:2]: praise: hey, nice, we matched up in how we exposed our user store :)

File: src/lib/features/auth/specs/auth.spec.ts [hunk 2]
```diff
@@ -0,0 +1,17 @@
+import { get } from 'svelte/store';
```
Comment (illright) [inline src/lib/features/auth/specs/auth.spec.ts:1]: hate to tell you this, but this won't be necessary :p. I implemented it in #126 (which you can review, btw)

File: src/lib/features/auth/specs/login-button.spec.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import svelte from 'svelte-inline-compile';
+import { LoginButton } from '..';
+import { useLocalVars } from '$lib/shared/lib/jest-hacks';
+import { render } from '@testing-library/svelte';
+
+it('renders the button', async () => {
```
Comment (illright) [inline src/lib/features/auth/specs/login-button.spec.ts:6]: issue: I think it's important for the user that the button is clickable and actually calls the authentication flow on press, which is not covered by the tests currently

File: src/lib/features/auth/specs/login-button.spec.ts [hunk 4]
```diff
@@ -0,0 +1,11 @@
...
+  const loginButton = useLocalVars(svelte`<LoginButton />`, [LoginButton]);
```
Comment (illright) [inline src/lib/features/auth/specs/login-button.spec.ts:7]: suggestion: no need for `svelte-inline-compile` here, you can just pass the `LoginButton` straight to the `render()` function

File: src/lib/pages/library/ui/library-page.svelte [hunk 5]
```diff
@@ -2,12 +2,22 @@
   import { H1 } from '$lib/shared/ui';
   import { likesStore } from '$lib/features/like-episode';
   import { EpisodeCard } from '$lib/widgets/episode';
+  import { LoginContainer } from '$lib/widgets/user';
```
Comment (illright) [inline src/lib/pages/library/ui/library-page.svelte:5]: issue: it sounds strange to me how a "login container" is inside a widget called "user":
* I don't really understand what is a login container
* I don't understand what could the "user" widget contain

suggestion: let's move the code that represents the user into `entities/user` and combine it with the buttons on the `pages` layer, I don't think we need widgets here

File: src/lib/widgets/user/index.ts [hunk 6]
```diff
@@ -0,0 +1 @@
+export { default as LoginContainer } from './ui/login.svelte';
```
Comment (illright) [inline src/lib/widgets/user/index.ts:1]: suggestion (non-blocking): naming components exactly after their filename helps to navigate around the project, I suggest we call this file `login-container` (although, according to my previous comments, I don't think we need this file at all)

File: src/lib/widgets/user/ui/avatar.svelte [hunk 7]
```diff
@@ -0,0 +1,7 @@
+<script lang="ts">
+  export let image: string;
+</script>
+
+<div class="aspect-square h-12 w-12 rounded-full overflow-hidden">
+  <img src={image} alt="user avatar" />
```
Comment (illright) [inline src/lib/widgets/user/ui/avatar.svelte:6]: issue (non-blocking): I'm not sure if this alt text is actually helpful. Alt text works best when it describes the content of an image, not its purpose, which is usually obvious from context

**Review by illright — CHANGES_REQUESTED** [review]
I just submitted my PR on the authentication flow, it may be helpful for you to get it merged to avoid writing placeholders in place of authentication logic. So you're welcome to review #126 to get it merged faster :)

File: src/lib/features/auth/specs/login-button.spec.ts (hunk 3 above)
Comment (VanishMax) [reply src/lib/features/auth/specs/login-button.spec.ts:6]: I would really like if you provide an idea on how to actually do it

Comment (illright) [reply src/lib/features/auth/specs/login-button.spec.ts:6]: You need to mock the `$lib/features/authenticate` module, trigger a click on a button with `@testing-library/user-event` and then check the mock to see if it has been called or not

**Comment by VanishMax** [comment]
@illright I merged your PR into this one and connected UI with the logic. Everything's good except one thing: when I log in using GitHub, it redirects me back to the website, but the `user` writable store becomes filled only after reloading the page

File: src/lib/features/authenticate/ui/specs/sign-in-button.spec.ts [hunk 8]
```diff
@@ -0,0 +1,7 @@
+import SignInButton from '../sign-in-button.svelte';
+import { render } from '@testing-library/svelte';
+
+it('renders the button', async () => {
+  const { getByRole } = render(SignInButton);
+  expect(getByRole('button', { name: 'Sign in with GitHub' })).toBeInTheDocument();
```
Comment (illright) [inline src/lib/features/authenticate/ui/specs/sign-in-button.spec.ts:6]: question: so what about those mocks that we discussed above?

**Review by illright — COMMENTED** [review]
issue (non-blocking): you didn't delete the `AudioFetch` file, you made it empty. I deleted it in my PR, so now this PR adds it back again

**Comment by illright** [comment]
> but the user writable store becomes filled only after reloading the page

That's weird, that should not happen. Oh well, let's keep it this way for now and address this problem in #127

File: src/lib/features/authenticate/ui/specs/sign-out-button.spec.ts [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+  signOut: jest.fn(),
+}));
+
+import { signOut } from '../../model/sign-out';
+
+it('renders and clicks the sign out button', async () => {
+  const user = userEvent.setup();
+  const { getByRole } = render(SignOutButton);
+  await user.click(getByRole('button', { name: 'Sign out' }));
+  expect(signOut).toHaveBeenCalledTimes(1);
```
Comment (illright) [inline src/lib/features/authenticate/ui/specs/sign-out-button.spec.ts:15]: nitpick (non-blocking): I think the number of times this function has been called is not that important in this case, there's a shorter `.toHaveBeenCalled()` assertion. But this works too, and asserts even more things

File: src/lib/features/authenticate/ui/specs/sign-out-button.spec.ts [hunk 10]
```diff
@@ -0,0 +1,16 @@
...
+import { render } from '@testing-library/svelte';
+import userEvent from '@testing-library/user-event';
+
+jest.mock('../../model/sign-out', () => ({
+  signOut: jest.fn(),
+}));
+
+import { signOut } from '../../model/sign-out';
+
+it('renders and clicks the sign out button', async () => {
```
Comment (illright) [inline src/lib/features/authenticate/ui/specs/sign-out-button.spec.ts:11]: nitpick (non-blocking): I like to read these `it`-s as being about the filename. So saying that a "sign-out-button" renders and clicks the sign out button feels strange to me. I'd phrase it as a "sign-out-button" calls the `signOut` function on click

File: src/lib/features/authenticate/ui/specs/sign-out-button.spec.ts [hunk 11]
```diff
@@ -0,0 +1,16 @@
+import SignOutButton from '../sign-out-button.svelte';
+import { render } from '@testing-library/svelte';
+import userEvent from '@testing-library/user-event';
+
+jest.mock('../../model/sign-out', () => ({
+  signOut: jest.fn(),
+}));
+
+import { signOut } from '../../model/sign-out';
```
Comment (illright) [inline src/lib/features/authenticate/ui/specs/sign-out-button.spec.ts:9]: thoughts (non-blocking): by the way, Jest hoists the `jest.mock` calls so you don't actually have to import after the mock. But I see why you'd want to do this

**Review by illright — APPROVED** [review]
praise: the tests are good! Clean and accessible

Not sure what's wrong with the Cloudflare preview, it showed me the button once, I logged in, then everything related to `UserContainer` has just disappeared.

**Comment by illright** [comment]
Oh, I understood what's wrong with the Cloudflare demo. I logged in from the preview deployment, but it redirected me to the stable deployment. That's a shame, but whatever

**Pull request merged**